dataset/src/main/scala/frameless/TypedEncoder.scala

```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
     }
   }
 
+  implicit def arrayEncoder[A : ClassTag](
+    implicit
+    underlying: TypedEncoder[A]
+  ): TypedEncoder[Array[A]] = new CollectionEncoder[Array, A]() {
+    val nullable: Boolean = false
+
+    val sourceDataType: DataType = FramelessInternals.objectTypeFor[Array[A]](classTag)
+
+    val targetDataType: DataType = DataTypes.createArrayType(underlying.targetDataType)
+
+    def constructorFor(path: Expression): Expression = arrayData(path)
+
+    def extractorFor(path: Expression): Expression = {
+      // if source `path` is already native for Spark, no need to `map`
+      if (ScalaReflection.isNativeType(underlying.sourceDataType)) {
+        NewInstance(
+          classOf[GenericArrayData],
+          path :: Nil,
+          dataType = ArrayType(underlying.targetDataType, underlying.nullable)
+        )
+      } else {
+        MapObjects(underlying.extractorFor, path, underlying.sourceDataType)
+      }
+    }
+  }
+
+  implicit def mapEncoder[A : ClassTag, B : ClassTag](
+    implicit
+    encodeA: TypedEncoder[A],
+    encodeB: TypedEncoder[B]
+  ): TypedEncoder[Map[A, B]] = new TypedEncoder[Map[A, B]] {
+    val nullable: Boolean = false
+    val sourceDataType = FramelessInternals.objectTypeFor[Map[A, B]]
+    val targetDataType = MapType(encodeA.targetDataType, encodeB.targetDataType, encodeB.nullable)
+
+    // needed to construct Map using `ArrayBasedMapData.toScalaMap` (avoids an extra MapObjects)
+    private val arrayA = arrayEncoder[A]
```

Comment:
Actually, I think it might be better to capture them as vals.

Imagine `TypedDataset[Array[Map[Int, Int]]` - the outer encoder is going to call the inner encoder potentially thousands or millions of times during a plan execution. If we call `arrayEncoder[A]` and `arrayEncoder[B]` each time, it is going to result in a lot of needless allocation/dellocation and will potentially hurt hotspotting (since each instance of `arrayEncoder[A]` output is its own refinement class - maybe that should be changed?).

Does capturing the values do harm?